### PR TITLE
WIP - Postgres Connector

### PR DESCRIPTION
**Pull request by ejoebstl**
This one is WIP, but I would need help/feedback, especially for error handling. Please see the comments.

### CODE REVIEW

File: src/connector/postgres.rs [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+                },
+                PostgresType::UUID => match row.try_get(i)? {
+                    Some(val) => {
+                        let val: Uuid = val;
+                        ParameterizedValue::Uuid(val)
+                    }
+                    None => ParameterizedValue::Null,
+                },
+                /*
+                TODO: How to support arrays?
```
Comment (ejoebstl) [inline src/connector/postgres.rs]: How should we handle Array types?
A) Add to ParameterizedValue
B) Don't handle at all

File: src/connector/postgres.rs [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+        let mut client = self.pool.get()?;
+        let result = f(&mut client);
+        result
+    }
+}
+
+impl From<postgres::error::Error> for Error {
+    fn from(e: postgres::error::Error) -> Error {
+        dbg!(&e);
+        // TODO: Ask J how to map to Failure::Error
```
Comment (ejoebstl) [inline src/connector/postgres.rs]: Error Handling: postgres::error::Error does not use the failure crate. What would be a good way to convert here?

File: src/connector/postgres.rs [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+        Error::NotFound
+    }
+}
+
+impl From<native_tls::Error> for Error {
+    fn from(e: native_tls::Error) -> Error {
+        dbg!(&e);
+        // TODO: Ask J how to map to Failure::Error
```
Comment (ejoebstl) [inline src/connector/postgres.rs]: Error Handling: postgres::error::Error does not use the failure crate. What would be a good way to convert here?

File: Cargo.toml [hunk 4]
```diff
@@ -4,19 +4,21 @@ version = "0.1.0"
...
-default = ["rusqlite-client", "json-1", "postgresql-0_16", "uuid-0_7", "chrono-0_4", "mysql_sync"]
+default = ["rusqlite-client", "json-1", "postgresql-0_16", "uuid-0_7", "chrono-0_4", "mysql_sync", "postgresql" ]
 rusqlite-client = ["rusqlite", "r2d2_sqlite", "r2d2", "libsqlite3-sys"]
 sqlite3 = ["sqlite", "r2d2", "r2d2_sqlite"]
 json-1 = ["serde_json"]
 postgresql-0_16 = ["postgres", "rust_decimal", "r2d2"]
 uuid-0_7 = ["uuid"]
 chrono-0_4 = ["chrono"]
 mysql_sync = ["mysql", "r2d2"]
+postgresql = ["postgres", "r2d2_postgres", "native-tls", "rust_decimal", "tokio-postgres", "tokio-postgres-native-tls"]
```
Comment (pimeys) [inline Cargo.toml]: Use the already existing feature!

File: src/connector/postgres.rs (hunk 1 above)
Comment (pimeys) [reply src/connector/postgres.rs]: A.

File: src/connector/postgres.rs (hunk 2 above)
Comment (pimeys) [reply src/connector/postgres.rs]: https://github.com/prisma/prisma/blob/alpha/server/prisma-rs/query-engine/connectors/sql-connector/src/error.rs#L242

The Failure crate provides the needed `From`.

File: src/connector/postgres.rs (hunk 3 above)
Comment (pimeys) [reply src/connector/postgres.rs]: https://github.com/prisma/prisma/blob/alpha/server/prisma-rs/query-engine/connectors/sql-connector/src/error.rs#L242

File: tests/postgres_connector.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
+use postgres::Config;
```
Comment (pimeys) [inline tests/postgres_connector.rs]: You should write the tests to the postgres connector.

Comment (ejoebstl) [reply tests/postgres_connector.rs]: Aka I should put them in the same file?

Comment (pimeys) [reply tests/postgres_connector.rs]: No, write the tests to the code file you're testing. In this case, after the code, add

```rust
#[cfg(test)]
mod tests {
...
}
```

File: src/connector/postgres.rs (hunk 1 above)
Comment (ejoebstl) [reply src/connector/postgres.rs]: Done.

File: src/connector/postgres.rs (hunk 2 above)
Comment (ejoebstl) [reply src/connector/postgres.rs]: Done.

File: src/connector/postgres.rs (hunk 3 above)
Comment (ejoebstl) [reply src/connector/postgres.rs]: Done

File: Cargo.toml (hunk 4 above)
Comment (ejoebstl) [reply Cargo.toml]: postgresql-0_16?

File: tests/postgres_connector.rs (hunk 5 above)
Comment (ejoebstl) [reply tests/postgres_connector.rs]: Could we postpone that after prisma day? Right now we have three variants in prisma: test folder, doc test, test in the same file. 

Let's set this when we have more time. :)

File: Cargo.toml (hunk 4 above)
Comment (ejoebstl) [reply Cargo.toml]: Done.

File: src/visitor/sqlite.rs [hunk 6]
```diff
@@ -118,7 +118,7 @@ impl Bindable for ParameterizedValue {
     }
 }
 
-// TODO: This most likely needs to be in another class.
+// TODO: This most likely should be in another class, as it is not related to sqlite.
```
Comment (pimeys) [inline src/visitor/sqlite.rs]: Um... it is kind of related to sqlite...

Comment (ejoebstl) [reply src/visitor/sqlite.rs]: My bad, looked in the wrong namespace.

**Comment by ejoebstl** [comment]
Array feature is now included in the postgresql-xx feature.

**Pull request merged**